routes/callback_discord.go

```diff
@@ -0,0 +1,73 @@
+package routes
+
+import (
+	"encoding/json"
+	"net/http"
+	"os"
+
+	"github.com/bwmarrin/discordgo"
+	"github.com/code-golf/code-golf/session"
+	"golang.org/x/oauth2"
+)
+
+var discordConfig = oauth2.Config{
+	ClientID:     os.Getenv("DISCORD_CLIENT_ID"),
+	ClientSecret: os.Getenv("DISCORD_CLIENT_SECRET"),
+	Endpoint: oauth2.Endpoint{
+		AuthURL:   discordgo.EndpointOauth2 + "authorize",
+		TokenURL:  discordgo.EndpointOauth2 + "token",
+		AuthStyle: oauth2.AuthStyleInParams,
+	},
+}
+
+// CallbackDiscord serves GET /callback/discord
+func CallbackDiscord(w http.ResponseWriter, r *http.Request) {
+	if r.FormValue("code") == "" {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	golfer := session.Golfer(r)
+	if golfer == nil {
+		w.WriteHeader(http.StatusUnauthorized)
+		return
+	}
+
+	discordConfig.RedirectURL = "https://" + r.Host + "/callback/discord"
```

Comment:
I'm not sure if this is racey, since discordConfig is a global. Granted we always set it to the same value. Really ought to get some tests going with `-race` at some point.

We do know what it ought to be at init time, the simple way would to make it conditional on the dev flag, ideally we'd have a config.Host or something.